cmd/zoekt-sourcegraph-indexserver/sg.go

```diff
@@ -85,16 +87,40 @@ type Sourcegraph interface {
 	UpdateIndexStatus(repositories []indexStatus) error
 }
 
-func newSourcegraphClient(rootURL *url.URL, hostname string, batchSize int) *sourcegraphClient {
+type SourcegraphClientOption func(*sourcegraphClient)
 
-	client := retryablehttp.NewClient()
-	client.Logger = debug
+// WithBatchSize controls how many repository configurations we request a time.
+// If BatchSize is 0, we default to requesting 10,000 repositories at once.
+func WithBatchSize(batchSize int) SourcegraphClientOption {
+	return func(c *sourcegraphClient) {
+		c.BatchSize = batchSize
+	}
+}
+
+// WithShouldUseGRPCFunc sets the function that is used to determine whether to use gRPC
+// to communicate with Sourcegraph. This function must be thread-safe.
+func WithShouldUseGRPCFunc(shouldUseGRPCFunc func() bool) SourcegraphClientOption {
+	return func(c *sourcegraphClient) {
+		c.shouldUseGRPCFunc = shouldUseGRPCFunc
+	}
+}
+
+// WithGRPCClient sets the gRPC client to use for communicating with Sourcegraph.
+func WithGRPCClient(client proto.IndexedSearchConfigurationServiceClient) SourcegraphClientOption {
```

Comment:
One aspect of this interface that isn't ideal is that it's a bit of "hidden knowledge" that you have to both provide a custom `WithShouldUseGRPCFunc` and Provide a `WithUseGRPCClient` to actually enable GRPC.

I guess this will be less problematic when gRPC is enabled by default (eliminating the need for WithShouldUseGRPCFunc), but I'd appreciate any suggestions here.

Reply:
Discussed offline. We've decided that continuing to provide a separate (use_grpc) option along with grpcClient option is the best path forward.